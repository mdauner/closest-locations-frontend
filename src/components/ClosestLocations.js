import React, { useState } from 'react'
import queryString from 'query-string'
import Loader from 'react-loader-spinner'
import LocationItem from './LocationItem'
import './ClosestLocations.scss'

const API_BASE_URL = process.env.REACT_APP_API_BASE_URL

function ClosestLocations() {
  const [loading, setLoading] = useState(false)
  const [locationNames, setLocationNames] = useState([])
  const [locations, setLocations] = useState([])
  const [error, setError] = useState('')

  async function fetchClosestLocations() {
    if (locationNames.length > 0) {
      setLoading(true)
      const response = await fetch(
        `${API_BASE_URL}/closest-locations?${queryString.stringify({
          locations: locationNames
        })}`
      )
      const responseText = await response.text()
      try {
        const newLocations = JSON.parse(responseText)
        setLocations(newLocations)
        setError('')
      } catch (e) {
        setLocations([])
        setError(responseText)
      }
      setLoading(false)
    } else {
      setLocations([])
    }
  }

  return (
    <div className="closest-locations">
      <h3>Closest Location Calculator</h3>
      <textarea
        className="closest-locations__textarea"
        placeholder="Please enter locations, separated by linebreaks."
        value={locationNames.join('\n')}
        onChange={e => setLocationNames(e.target.value.split('\n'))}
      />
      <button
        className="closest-locations__fetch-button"
        type="button"
        disabled={locationNames.length < 2}
        onClick={fetchClosestLocations}
      >
        Get Closest Locations
      </button>
      {!loading && error && <div className="error">Error: {error}</div>}
      {loading ? (
        <div className="spinner">
          <Loader
            type="Puff"
            color="hsl(120, 40%, 50%)"
            height="50"
            width="50"
          />
        </div>
      ) : (
        <div className="closest-locations__locations">
          {locations.map(LocationItem)}
        </div>
      )}
    </div>
  )
}

export default ClosestLocations
