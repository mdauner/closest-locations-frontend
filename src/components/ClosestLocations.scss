.closest-locations {
  width: 100%;
  padding: 2rem;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  align-items: center;

  &__textarea {
    width: 100%;
    height: 200px;
    padding: 1rem;
    box-sizing: border-box;
  }

  &__fetch-button {
    width: 100%;
    margin: 1rem 0;
    background-color: hsl(120, 40%, 50%);
    border: 2px solid hsl(120, 40%, 40%);
    color: white;
    font-size: 1.2rem;
    padding: 1rem;

    &:hover,
    &:focus {
      background-color: hsl(120, 40%, 60%);
      border: 2px solid hsl(120, 40%, 50%);
      cursor: pointer;
    }

    &[disabled] {
      background-color: hsl(120, 40%, 70%);
      border: 2px solid hsl(120, 40%, 60%);
      cursor: default;
    }
  }

  &__locations {
    width: 100%;
  }
}

.spinner {
  margin: 2rem;
}

.error {
  background-color: hsl(0, 50%, 70%);
  border: 1px solid hsl(0, 50%, 60%);
  color: white;
  padding: .5rem;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
}