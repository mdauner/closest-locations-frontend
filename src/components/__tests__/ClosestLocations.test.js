import React from 'react'
import { render, fireEvent, waitForElement } from 'react-testing-library'
import ClosestLocations from '../ClosestLocations'

beforeEach(() => {
  fetch.resetMocks()
})

test('Renders ClosestLocations correctly', async () => {
  const { container } = render(<ClosestLocations />)
  expect(container.firstChild).toMatchSnapshot()
})

test('Enables Button when more than two lines are entered in textarea', () => {
  const { getByText, getByPlaceholderText } = render(<ClosestLocations />)
  const textarea = getByPlaceholderText(/enter locations/i)
  const fetchButton = getByText(/get closest locations/i)
  fireEvent.change(textarea, {
    target: { value: 'New York\nSan Francisco' }
  })
  expect(fetchButton).not.toHaveAttribute('disabled')
})

test('Shows error message when less than two locations are valid', async () => {
  fetch.mockResponseOnce('error message')

  const { getByText, getByPlaceholderText } = render(<ClosestLocations />)
  const textarea = getByPlaceholderText(/enter locations/i)
  const fetchButton = getByText(/get closest locations/i)
  fireEvent.change(textarea, {
    target: { value: 'New York\nasdfasdf' }
  })
  fireEvent.click(fetchButton)
  await waitForElement(() => getByText(/error/i))
})

test('Shows location items', async () => {
  fetch.mockResponseOnce(
    JSON.stringify([
      {
        closestLocation: 'Paris',
        distance: 878606,
        location: 'Berlin'
      },
      {
        closestLocation: 'Berlin',
        distance: 878606,
        location: 'Paris'
      },
      {
        closestLocation: 'Paris',
        distance: 5857061,
        location: 'The Statue of Liberty'
      }
    ])
  )

  const { getByText, getByPlaceholderText, getAllByTestId } = render(
    <ClosestLocations />
  )
  const textarea = getByPlaceholderText(/enter locations/i)
  const fetchButton = getByText(/get closest locations/i)
  fireEvent.change(textarea, {
    target: { value: 'New York\nParis\nThe Statue of Liberty' }
  })
  fireEvent.click(fetchButton)
  const locations = await waitForElement(() => getAllByTestId('location-item'))
  expect(locations.length).toEqual(3)
})
