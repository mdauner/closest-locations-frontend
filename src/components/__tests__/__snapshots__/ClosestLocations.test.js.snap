// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ClosestLocations correctly 1`] = `
<div
  class="closest-locations"
>
  <h3>
    Closest Location Calculator
  </h3>
  <textarea
    class="closest-locations__textarea"
    placeholder="Please enter locations, separated by linebreaks."
  />
  <button
    class="closest-locations__fetch-button"
    disabled=""
    type="button"
  >
    Get Closest Locations
  </button>
  
  <div
    class="closest-locations__locations"
  />
</div>
`;
